### PR TITLE
fix: breadcrumb cache issue

### PR DESCRIPTION
**Pull request by AkshayBandi027**
## What does this PR do?

Fixes: #2193 

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Chore (refactoring code, technical debt, workflow improvements)
- [ ] Enhancement (small improvements)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How should this be tested?

-Go to dashboard/api/[apiId]/settings
- Change the name of api, see reflects the breadcrum immediately. 

### Required

- [x] Filled out the "How to test" section in this PR
- [x] Read [Contributing Guide](./CONTRIBUTING.md)
- [x] Self-reviewed my own code
- [ ] Commented on my code in hard-to-understand areas
- [x] Ran `pnpm build`
- [x] Ran `pnpm fmt`
- [x] Checked for warnings, there are none
- [x] Removed all `console.logs`
- [x] Merged the latest changes from main onto my branch with `git pull origin main`
- [x] My changes don't cause any responsiveness issues

### Appreciated

- [x] If a UI change was made: Added a screen recording or screenshots to this PR
https://www.loom.com/share/e31346feb61645c3a9ce31de667ce7c7?sid=bfddd2d0-8745-489c-8810-69e86ce1379f

## Summary by CodeRabbit

- **New Features**
	- Enhanced caching mechanisms for various components, improving data retrieval efficiency and cache management.
	- Added search functionality within the glossary terms through the `FilterableCommand` component.

- **Bug Fixes**
	- Improved error handling by redirecting users when a namespace is not found.

- **Documentation**
	- Updated template list showcasing the use of Unkey with various frameworks, providing additional resources for developers.

### CODE REVIEW

File: apps/dashboard/app/(app)/@breadcrumb/authorization/roles/[roleId]/page.tsx [hunk 1]
```diff
@@ -25,6 +25,8 @@ async function AsyncPageBreadcrumb(props: PageProps) {
       await db.query.roles.findFirst({
         where: (table, { eq }) => eq(table.id, roleId),
       }),
+    ["roleById"],
+    { tags: [`role-${props.params.roleId}`] },
```
[the remark this answers is not in the document]
Comment (AkshayBandi027) [reply apps/dashboard/app/(app)/@breadcrumb/authorization/roles/[roleId]/page.tsx:29]:
```suggestion
    ["roleById"],
    { tags: [`role-${props.params.roleId}`] }, 
   
```
please check authorization/permissions/roles/[roleId]/update-role.tsx
we are revalidating the tag!

File: apps/dashboard/app/(app)/authorization/permissions/[permissionId]/client.tsx [hunk 2]
```diff
@@ -1,5 +1,6 @@
 "use client";
 
+import { revalidateMyTag } from "@/app/(app)/apis/[apiId]/settings/actions";
```
[the remark this answers is not in the document]
Comment (AkshayBandi027) [reply apps/dashboard/app/(app)/authorization/permissions/[permissionId]/client.tsx:3]: Should I make specific actions.ts file for dashboard that can be reused across the dashboard ! Leave as it is!

File: apps/dashboard/app/(app)/apis/[apiId]/settings/actions.ts [hunk 3]
```diff
@@ -1,7 +1,11 @@
 "use server";
 
-import { revalidatePath } from "next/cache";
+import { revalidatePath, revalidateTag } from "next/cache";
 
 export async function revalidate() {
   await revalidatePath("/", "layout");
 }
+
+export async function revalidateMyTag(tag: string) {
```
Comment (chronark) [inline apps/dashboard/app/(app)/apis/[apiId]/settings/actions.ts:9]: what's the purpose of this function?
why don't we just import `revalidateTag` from `next/cache` in our components?

**Review by chronark — COMMENTED** [review]
I wonder if it's worth it to create a typesafe wrapper for cache tags

The problem right now is that a simple typo in one of the tags can cause really hard to debug problems.

wdyt?

File: apps/dashboard/app/(app)/apis/[apiId]/settings/actions.ts (hunk 3 above)
Comment (AkshayBandi027) [reply apps/dashboard/app/(app)/apis/[apiId]/settings/actions.ts:9]: we can't use revalidateTag in client side, we are creating wrapper around to use it in client side !

Comment (chronark) [reply apps/dashboard/app/(app)/apis/[apiId]/settings/actions.ts:9]: aah gotcha, makes sense

but then let's reexport it as the same name, `revalidateMyTag` looks weird

Comment (AkshayBandi027) [reply apps/dashboard/app/(app)/apis/[apiId]/settings/actions.ts:9]: "Agreed, a typesafe wrapper for cache tags would help prevent typos and make debugging easier.

Comment (AkshayBandi027) [reply apps/dashboard/app/(app)/apis/[apiId]/settings/actions.ts:9]: "Should we make any other changes or does everything look fine to you?"


> aah gotcha, makes sense
> 
> but then let's reexport it as the same name, `revalidateMyTag` looks weird

Ok ! "Should we make any other changes or does everything look fine to you?"

**Comment by AkshayBandi027** [comment]
> I wonder if it's worth it to create a typesafe wrapper for cache tags
> 
> The problem right now is that a simple typo in one of the tags can cause really hard to debug problems.
> 
> wdyt?

@chronark,  what think of this approach 

`export const createCacheKeyForSolutions = (slug: string) => `challenge-${slug}-solutions`;
export const createCacheKeyForSharedSolutionsTab = (userId: string) => `${userId}-shared-solutions`;

`revalidateTag(createCacheKeyForSolutions(slug));`
`

In **TYPEHERO** they these to create cache key to revalidated by tags

**Comment by perkinsjr** [comment]
> `export const createCacheKeyForSolutions = (slug: string) => `challenge-${slug}-solutions`; export const createCacheKeyForSharedSolutionsTab = (userId: string) => `${userId}-shared-solutions`;

This is fine approach. 

I trust that Trashh knows what he is doing.

**Comment by AkshayBandi027** [comment]
> > > I wonder if it's worth it to create a typesafe wrapper for cache tags
> > > The problem right now is that a simple typo in one of the tags can cause really hard to debug problems.
> > > wdyt?
> > 
> > 
> > @chronark, what do you think of this approach
> > `export const createCacheKeyForSolutions = (slug: string) => `challenge-${slug}-solutions`; export const createCacheKeyForSharedSolutionsTab = (userId: string) => `${userId}-shared-solutions`;
> > `revalidateTag(createCacheKeyForSolutions(slug));` ` https://github.com/typehero/typehero/blob/main/apps/web/src/app/challenge/%5Bslug%5D/solutions/_components/_actions.ts
> > In **TYPEHERO** they these to create cache key to revalidated by tags
> 
> why do you prefer functions over string literals?
> 
> ```ts
> type CacheKeyApi = `api-${string}`
> type CacheKeyPermission = `permission-${string}`
> 
> type CacheKey = CacheKeyApi | CacheKeyPermission | ...
> 
> export function revalidate(key CacheKey) {
> ...
> }
> ```

I  think   both approach's are type-safe and in **functions**  we hardcoding the cache key permissions , api and provide the slug, in contrast to second approach we using string literal to create union type  and passing key that is of type union type it to the revalidateTag !

I personally I don't see much difference ! If you feel **string literal with union type** is way to go then sure !

**Comment by chronark** [comment]
fair,

can we group it into a single export though, that'll make it much cleaner to import and use 

something like:
``` export const tags = {
  api: (apiId: string): string => {
    return `api-${apiId}`
  },
  ...
}

then you can use it as `{ tags: [tags.api(apiId) ]}`

File: apps/dashboard/app/(app)/ratelimits/page.tsx [hunk 4]
```diff
@@ -37,6 +37,8 @@ export default async function RatelimitOverviewPage() {
     return redirect("/new");
   }
 
+  console.log(workspace.ratelimitNamespaces,`workspace.ratelimitNamespaces`)
```
Comment (chronark) [inline apps/dashboard/app/(app)/ratelimits/page.tsx:40]: please remove

File: apps/dashboard/app/(app)/authorization/permissions/page.tsx [hunk 5]
```diff
@@ -43,6 +43,7 @@ export default async function RolesPage() {
     return redirect("/new");
   }
 
+  console.log(workspace.permissions, `workspace.permissions`)
```
Comment (chronark) [inline apps/dashboard/app/(app)/authorization/permissions/page.tsx:46]: remove

File: apps/dashboard/app/(app)/@breadcrumb/tags.ts [hunk 6]
```diff
@@ -0,0 +1,7 @@
+
+export const tags = {
+    api: (apiId: string): string => `api-${apiId}`,
```
Comment (chronark) [inline apps/dashboard/app/(app)/@breadcrumb/tags.ts:3]: these are not only relevant to breakdcrumbs let's move them to /app/lib/cache/tags.ts and also change all of the imports to `@/lib/cache`

**Review by chronark — CHANGES_REQUESTED** [review]

**Review by chronark — APPROVED** [review]

**Pull request merged**